// @google/adk/dist/cjs/code_executors/built_in_code_executor.js
/**
  * @license
  * Copyright 2025 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var built_in_code_executor_exports = {};
__export(built_in_code_executor_exports, {
  BuiltInCodeExecutor: () => BuiltInCodeExecutor,
  isBuiltInCodeExecutor: () => isBuiltInCodeExecutor
});
module.exports = __toCommonJS(built_in_code_executor_exports);
var import_model_name = require("../utils/model_name.js");
var import_base_code_executor = require("./base_code_executor.js");
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
var _a, _b;
const BUILT_IN_CODE_EXECUTOR_SIGNATURE_SYMBOL = Symbol.for("google.adk.builtInCodeExecutor");
function isBuiltInCodeExecutor(obj) {
  return typeof obj === "object" && obj !== null && BUILT_IN_CODE_EXECUTOR_SIGNATURE_SYMBOL in obj && obj[BUILT_IN_CODE_EXECUTOR_SIGNATURE_SYMBOL] === true;
}
class BuiltInCodeExecutor extends (_b = import_base_code_executor.BaseCodeExecutor, _a = BUILT_IN_CODE_EXECUTOR_SIGNATURE_SYMBOL, _b) {
  constructor() {
    super(...arguments);
    /** A unique symbol to identify BuiltInCodeExecutor class. */
    this[_a] = true;
  }
  executeCode(params) {
    return Promise.resolve({
      stdout: "",
      stderr: "",
      outputFiles: []
    });
  }
  processLlmRequest(llmRequest) {
    if (llmRequest.model && (0, import_model_name.isGemini2OrAbove)(llmRequest.model)) {
      llmRequest.config = llmRequest.config || {};
      llmRequest.config.tools = llmRequest.config.tools || [];
      llmRequest.config.tools.push({ codeExecution: {} });
      return;
    }
    throw new Error(`Gemini code execution tool is not supported for model ${llmRequest.model}`);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BuiltInCodeExecutor,
  isBuiltInCodeExecutor
});
