// @google/adk/dist/esm/tools/function_tool.js
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
var _a, _b;
import { Type } from "@google/genai";
import { ZodObject } from "zod";
import { isZodObject, zodObjectToSchema } from "../utils/simple_zod_to_json.js";
import { BaseTool } from "./base_tool.js";
function toSchema(parameters) {
  if (parameters === void 0) {
    return { type: Type.OBJECT, properties: {} };
  }
  if (isZodObject(parameters)) {
    return zodObjectToSchema(parameters);
  }
  return parameters;
}
const FUNCTION_TOOL_SIGNATURE_SYMBOL = Symbol.for("google.adk.functionTool");
function isFunctionTool(obj) {
  return typeof obj === "object" && obj !== null && FUNCTION_TOOL_SIGNATURE_SYMBOL in obj && obj[FUNCTION_TOOL_SIGNATURE_SYMBOL] === true;
}
class FunctionTool extends (_b = BaseTool, _a = FUNCTION_TOOL_SIGNATURE_SYMBOL, _b) {
  /**
   * The constructor acts as the user-friendly factory.
   * @param options The configuration for the tool.
   */
  constructor(options) {
    var _a2;
    const name = (_a2 = options.name) != null ? _a2 : options.execute.name;
    if (!name) {
      throw new Error(
        "Tool name cannot be empty. Either name the `execute` function or provide a `name`."
      );
    }
    super({
      name,
      description: options.description,
      isLongRunning: options.isLongRunning
    });
    /** A unique symbol to identify ADK function tool class. */
    this[_a] = true;
    this.execute = options.execute;
    this.parameters = options.parameters;
  }
  /**
   * Provide a schema for the function.
   */
  _getDeclaration() {
    return {
      name: this.name,
      description: this.description,
      parameters: toSchema(this.parameters)
    };
  }
  /**
   * Logic for running the tool.
   */
  async runAsync(req) {
    try {
      let validatedArgs = req.args;
      if (this.parameters instanceof ZodObject) {
        validatedArgs = this.parameters.parse(req.args);
      }
      return await this.execute(
        validatedArgs,
        req.toolContext
      );
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      throw new Error(`Error in tool '${this.name}': ${errorMessage}`);
    }
  }
}
export {
  FunctionTool,
  isFunctionTool
};
