// @google/adk/dist/esm/code_executors/built_in_code_executor.js
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
var _a, _b;
import { isGemini2OrAbove } from "../utils/model_name.js";
import { BaseCodeExecutor } from "./base_code_executor.js";
const BUILT_IN_CODE_EXECUTOR_SIGNATURE_SYMBOL = Symbol.for("google.adk.builtInCodeExecutor");
function isBuiltInCodeExecutor(obj) {
  return typeof obj === "object" && obj !== null && BUILT_IN_CODE_EXECUTOR_SIGNATURE_SYMBOL in obj && obj[BUILT_IN_CODE_EXECUTOR_SIGNATURE_SYMBOL] === true;
}
class BuiltInCodeExecutor extends (_b = BaseCodeExecutor, _a = BUILT_IN_CODE_EXECUTOR_SIGNATURE_SYMBOL, _b) {
  constructor() {
    super(...arguments);
    /** A unique symbol to identify BuiltInCodeExecutor class. */
    this[_a] = true;
  }
  executeCode(params) {
    return Promise.resolve({
      stdout: "",
      stderr: "",
      outputFiles: []
    });
  }
  processLlmRequest(llmRequest) {
    if (llmRequest.model && isGemini2OrAbove(llmRequest.model)) {
      llmRequest.config = llmRequest.config || {};
      llmRequest.config.tools = llmRequest.config.tools || [];
      llmRequest.config.tools.push({ codeExecution: {} });
      return;
    }
    throw new Error(`Gemini code execution tool is not supported for model ${llmRequest.model}`);
  }
}
export {
  BuiltInCodeExecutor,
  isBuiltInCodeExecutor
};
